src/Lean/Server/ServerTask.lean

```diff
@@ -0,0 +1,128 @@
+  - If the function being mapped is cheap, map it with `sync := true`. This runs the function on
+    the current thread if the task has already finished, or it reuses the thread of the task if the
+    task has not finished. This ensures that the function to be executed cannot be starved
+    by the elaborator.
+  - If the function being mapped is costly, map it with `prio := .dedicated`. This spawns a new
+    thread and thus cannot be starved by the elaborator.
+Finally, if the function being mapped is costly, but is already being executed in a dedicated task,
+it is fine to pretend that it is a cheap function instead.
+-/
+
+namespace Lean.Server
+
+structure ServerTask (α : Type u) where
+  task : Task α
+  deriving Inhabited
+
+instance : Coe (Task α) (ServerTask α) where
+  coe := .mk
+
+namespace ServerTask
+
+def pure (x : α) : ServerTask α := Task.pure x
+
+def get (t : ServerTask α) : α := t.task.get
+
+def mapCheap (f : α → β) (t : ServerTask α) : ServerTask β :=
+  t.task.map f (sync := true)
+
+def mapCostly (f : α → β) (t : ServerTask α) : ServerTask β :=
+  t.task.map f (prio := .dedicated)
+
+def bindCheap (t : ServerTask α) (f : α → ServerTask β) : ServerTask β :=
+  t.task.bind (f · |>.task) (sync := true)
+
+def bindCostly (t : ServerTask α) (f : α → ServerTask β) : ServerTask β :=
+  t.task.bind (f · |>.task) (prio := .dedicated)
+
+namespace BaseIO
+
+def asTask (act : BaseIO α) : BaseIO (ServerTask α) :=
+  _root_.BaseIO.asTask act
+
+def mapTaskCheap (f : α → BaseIO β) (t : ServerTask α) : BaseIO (ServerTask β) :=
+  BaseIO.mapTask f t.task (sync := true)
+
+def mapTaskCostly (f : α → BaseIO β) (t : ServerTask α) : BaseIO (ServerTask β) :=
+  BaseIO.mapTask f t.task (prio := .dedicated)
+
+def bindTaskCheap (t : ServerTask α) (f : α → BaseIO (ServerTask β)) : BaseIO (ServerTask β) :=
+  BaseIO.bindTask t.task (ServerTask.task <$> f ·) (sync := true)
+
+def bindTaskCostly (t : ServerTask α) (f : α → BaseIO (ServerTask β)) : BaseIO (ServerTask β) :=
+  BaseIO.bindTask t.task (ServerTask.task <$> f ·) (prio := .dedicated)
+
+end BaseIO
+
+namespace EIO
+
+def asTask (act : EIO ε α) : BaseIO (ServerTask (Except ε α)) :=
+  _root_.EIO.asTask act
+
+def mapTaskCheap (f : α → EIO ε β) (t : ServerTask α) : BaseIO (ServerTask (Except ε β)) :=
+  EIO.mapTask f t.task (sync := true)
+
+def mapTaskCostly (f : α → EIO ε β) (t : ServerTask α) : BaseIO (ServerTask (Except ε β)) :=
+  EIO.mapTask f t.task (prio := .dedicated)
+
+def bindTaskCheap (t : ServerTask α) (f : α → EIO ε (ServerTask (Except ε β))) : BaseIO (ServerTask (Except ε β)) :=
+  EIO.bindTask t.task (ServerTask.task <$> f ·) (sync := true)
+
+def bindTaskCostly (t : ServerTask α) (f : α → EIO ε (ServerTask (Except ε β))) : BaseIO (ServerTask (Except ε β)) :=
+  EIO.bindTask t.task (ServerTask.task <$> f ·) (prio := .dedicated)
+
+end EIO
+
+namespace IO
+
+def asTask (act : IO α) : BaseIO (ServerTask (Except IO.Error α)) :=
+  _root_.IO.asTask act
```

Comment:
Should this be using `prio := .dedicated`?

Reply:
Absolutely.